src/js/libcs/Json.js

```diff
@@ -0,0 +1,223 @@
+    var res;
+    if (modifs.iterator !== undefined) {
+        let it = evaluate(modifs.iterator);
+        let iterTypes = ["key", "value", "pair"];
+        if (it.ctype === 'string' && iterTypes.includes(it.value)) {
+            iteratorType = it.value;
+        }
+    }
+    // iterate over key, value or pair
+    if (iteratorType === "value") {
+        for (let k in li) {
+            namespace.setvar(runVar, li[k]);
+            res = evaluate(fct);
+        }
+    } else if (iteratorType === "key") {
+        for (let k in li) {
+            namespace.setvar(runVar, General.string(k));
+            res = evaluate(fct);
+        }
+    } else { // pair
+        for (let k in li) {
+            namespace.setvar(runVar, Json._helper.GenJSONAtom(k, li[k]));
+            res = evaluate(fct);
+        }
+    }
+
+    return res;
+};
+
+Json._helper.niceprint = function(a, modifs, options) {
+    if (a.ctype === "JSON") {
+        return Json.niceprint(a, modifs, options);
+    }
+
+    if (options.printValidJSON) {
+        if (a.ctype === "number" && a.value.imag === 0) {
+            return a.value.real;
+        }
+        if ("boolean" === a.ctype) {
+            return a.value;
+        }
+        if (a.ctype === 'list') {
+            var erg = "[";
+            for (var i = 0; i < a.value.length; i++) {
+                erg = erg + Json._helper.niceprint(evaluate(a.value[i]));
+                if (i !== a.value.length - 1) {
+                    erg = erg + ', ';
+                }
+
+            }
+            return erg + "]";
+        }
+
+        return "\"" + String(niceprint(a)) + "\"";
+    }
+
+    return niceprint(a);
+};
+
+Json.Atomniceprint = function(el) {
+    return "\{" + el.value.key + ":" + niceprint(el.value.value) + "\}";
+};
+
+Json.niceprint = function(el, modifs, options) {
+    if (!options) {
+        options = {};
+        options.printValidJSON = false;
+        options.printedWarning = false;
+        // track depth
+        options.visitedMap = {};
+        options.visitedMap.level = 0;
+        options.visitedMap.maxLevel = 1000;
+        options.visitedMap.maxElVisit = 5000;
+
+        if (modifs) {
+            if (modifs.JSON) {
+                let jmodif = evaluate(modifs.JSON);
+                if (jmodif.ctype === "boolean") options.printValidJSON = jmodif.value;
+            }
+            if (modifs.maxDepth) {
+                let depth = evaluate(modifs.maxDepth);
+                if (depth.ctype === "number") options.visitedMap.maxLevel = depth.value.real;
+            }
+        }
+    }
+
+    var visitedMap = options.visitedMap;
+    // track a new recursive call
+    visitedMap.newLevel = true;
+    visitedMap.level += 1;
+
+    var keys = Object.keys(el.value).sort();
+    var jsonString = "{" + keys.map(function(key) {
+        // update visitedMap
+        let elValKey = el.value[key];
+        if (!visitedMap[elValKey]) {
+            visitedMap[elValKey] = 1;
```

Comment:
This line is equivalent to
`visitedMap["[object Object]"]=1;`.
Different key-value pair are not distinguished by this approach.
because CindyScript-values are objects and casted to a string when used as a key.

Maybe you just simplify this code and bound the DFS on the total number of visited nodes?